### ISSUE TITLE
Expose more raw Keyv functionality to enable better extensions.

### ISSUE THREAD

**Issue by lukechilds** [jaredwray/keyv#36]
If Keyv (and official storage adapters) were to expose more internals, this would allow for much richer functionality to be built on top of Keyv.

Mainly:

- [x] Optionally return the raw DB object (#39)
- [ ] Get all keys method (#40)
- [ ] Delete multiple keys in a single operation (#41)
- [ ] Get multiple keys in a single operation (#78)
- [ ] Check if key exists (#50 )

Main thing that comes to mind is a DB migrator (needs access to expiry dates) and [`keyv-shrink`](https://github.com/MySidesTheyAreGone/keyv) which is currently a fork because the desired functionality couldn't be implemented on top of Keyv.

With the three listed options implemented, `keyv-shrink` could be implemented efficiently on top of Keyv. All updates to Keyv and Keyv storage adapters would automatically work with `keyv-shrink`. Example of an implementation could be something like this:

```js
const Keyv = require('keyv');

class KeyvShrink extends Keyv {
  shrink() {
    ...
  }
}
```

@MySidesTheyAreGone What are your thoughts on this? I'm tied up on client project right now but I'm interested in adding this soon.

**Comment by MySidesTheyAreGone** [comment]
Yeah I didn't press this kind of things because you're obviously swamped but this is mirroring my thoughts very nicely. Kyev shouldn't be more complicated than it is, it should just be a bit more transparent.

Here's a rough timeline you can build on: I'm leaving next Saturday for a three weeks tour of India. That's about a month from now during which you and/or others can add some of that functionality without rushing it. When I'm back I can refactor -shrink and/or finish Keyvrator and we can get to the end of November with a significantly more powerful and complete Keyv.

**Comment by MySidesTheyAreGone** [comment]
(Or you can wait more and I can contribute the required changes to Keyv core of course - when I'm back)

**Comment by lukechilds** [comment]
All the changes should be 100% backwards compatible for Keyv, storage adapters and `keyv-shrink` so there's no rush to get this sorted, it won't affect users in any way.

Great, glad you're in agreement, I'm happy taking care of the Keyv side myself when I've got time, could maybe use a hand on some of the storage adapters once I've got it working in Keyv with tests.

**Comment by MySidesTheyAreGone** [comment]
Ok then, I will proceed when I get back! Thanks!

**Comment by skinofstars** [comment]
Hey both. Something that allows us to add more functionality to Keyv storage adapters would be great. In particular, I'm really hankering for a get all keys method.

**Comment by MySidesTheyAreGone** [comment]
Getting all the keys would be indeed useful for migrating data between keyvs.

(Cheers from Orchha, India)

**Comment by lukechilds** [comment]
@skinofstars Yeah this is 100% planned in. I actually just finished the client project that was draining a lot of my time yesterday.

So I'm gonna start working on this very soon (probably later this week).

@MySidesTheyAreGone Hope you're enjoying India!

**Comment by MySidesTheyAreGone** [comment]
It's an incredible trip. Halfway through now. Take care!

**Comment by Moeriki** [comment]
I'm also very interested in exposing `expires`.

If I can assist in implementing the changes do let me know.

**Comment by MySidesTheyAreGone** [comment]
Defining Symbol.iterator in Keyv's class:

```js
  * [Symbol.iterator] () {
    yield* this.opts.store;
  }
```

I can now do this:

```js
> var Keyv = require('./src/index.js')
> let test = new Keyv()
> test.set('one', 1)
> test.set('two', 2)
> test.set('three', 3)
> for (let item of test) {console.log(item)}
[ 'keyv:one', '{"value":1,"expires":null}' ]
[ 'keyv:two', '{"value":2,"expires":null}' ]
[ 'keyv:three', '{"value":3,"expires":null}' ]
undefined
```

This only works if the store is a Map, of course. To make it work on everything we need to implement [Symbol.iterator] in the other stores, which may or may not be difficult.

If we do make it work, it would be enough for a slower but compatible version of shrink() (right now it defines one more column) and keyvrator (which always goes through all of the keys anyway). Probably make a lot of other things possible too.

I don't know, food for thought.

**Comment by lukechilds** [comment]
I had a similar idea, I don't think having the Keyv class as iterable is a good idea because it won't work with async operations.

I was thinking adding `.keys()`, `.values()` `.entries()` methods similar to the `Map` spec but instead they would return a Promise that resolves to an iterable.

Also, they should just return the value by default. But all get/iterator methods should have the option to return raw data.

Something like:

```js
const keyv = new Keyv();
await keyv.set('foo', 'bar');
await keyv.set('fizz', 'buzz');

await keyv.get('foo');
// 'bar'
await keyv.get('foo', { raw: true });
// { value: 'bar', expires: null }

const entries = await keyv.entries();
for (const [key, value] of entries) {
  console.log(`${key} => ${value}`);
}
// 'foo => bar'
// 'fizz => buzz'

const entries = await keyv.entries({ raw: true });
for (const [key, value] of entries) {
  console.log(`${key} => ${JSON.stringify(value)}`);
}
// 'foo => {"value":"bar","expires":null}'
// 'fizz => {"value":"buzz","expires":null}'
```

Basically I wanna stay as close to the `Map` spec as possible while allowing for async operations.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map

**Comment by lukechilds** [comment]
Also

```js
await keyv.delete('foo', 'fizz');
```

Would be nice to have for performant deletion of multiple keys in a single operation. However not sure that's gonna be feasible to implement without breaking `Map` compatibility.

**Comment by MySidesTheyAreGone** [comment]
Yeah, ok:

> I was thinking adding .keys(), .values() .entries() methods similar to the Map spec but instead they would return a Promise that resolves to an iterable.

This is even more than I need, so no worries.

In the meantime enjoy a bit of science fiction (stage 3 proposal):

https://github.com/tc39/proposal-async-iteration

**Comment by MySidesTheyAreGone** [comment]
Dear diary,
today I was very bored.

keyv-sql (with keys() values() entries()):
https://gist.github.com/MySidesTheyAreGone/b44d31807460988bfa39e0d82dd842b8

keyv (with keys() values() entries() {raw:true} and a bit more):
https://gist.github.com/MySidesTheyAreGone/644350e1ae3ab1afe6ceb2f392838cd5

Not sure if useful, sure as hell saved me from a very dull day. Ignore if you're already working on it.

Quick test:
https://gist.github.com/MySidesTheyAreGone/24823e80048eb231447ea4261809b2fc

**Comment by lukechilds** [comment]
Nice!

Working on it but this may come in helpful, thanks!

**Comment by adityapatadia** [comment]
How can I help here to make get this issue closed?

**Comment by TehShrike** [comment]
A "Get all keys method" would increase the complexity of adapter implementations.  I believe that level of complexity is why so many [levelup](https://github.com/Level/levelup) storage implementations are languishing unmaintained.

If you have 50k keys, you can't fetch them all at once, so you need to expose some way to iterate over them.  Iterating over that many values in, say, a sql database, can be fiddly, especially if you want to try to guarantee that no keys are skipped or returned more than once.

I'm personally leaning towards keyv over levelup now because of the lower complexity level.

**Comment by adityapatadia** [comment]
As per my experience, I think it's always great to have following functionalities in Keyv.
- auto expiring keys and LRU cache - works great when keyv is used as cache
- Full API compatibility with Map - helps reduce memory growth, multiple workers can share same storage space and Map is already a good specification.

I don't think this would add huge complexities to codebase.

**Comment by lukechilds** [comment]
@TehShrike very valid points, check out https://github.com/lukechilds/keyv/issues/40 for more discussion on the points you bring up.

**Comment by Quantomistro3178** [comment]
Any more work done towards this? Couldn't you just expose some of the lower-level stuff from keyv-sql (or other modules), and allow users to run raw queries by themselves? It'd make it a lot more flexible.

**Comment by lukechilds** [comment]
It's not quite a simple as that.

The idea of Keyv is that any behaviour it exposes should be 100$ abstracted across all data stores.

If you're using the Keyv API you should be able to swap to any datastore backend and your code will still work. This is the main value proposition of Keyv.

Yes exposing features from underlying storage adapters would be more flexible but it would completely break this feature of Keyv.

We need to determine common usage requirements (The initial post in this issue) and find ways to abstract those into a common API that will provide consistent behaviour across all storage adapters.

Unfortunately I'm very busy with paid work at the moment so haven't had a huge amount of time to look into it.

**Comment by Quantomistro3178** [comment]
Ah, that's true. I chose to use Keyv for that reason as well at first, so it makes sense. Also I read your call for help, I'm not that experienced but would be willing to help out with whatever possible. I wouldn't want Keyv to go unmaintained, it's an awesome package. Also, good luck with your paid work lol 👍

**Comment by jaredwray** [comment]
closing this and moving to individual requests
